### ISSUE TITLE
Nice debug printing for error types

### ISSUE THREAD

**Issue by esmusick** [tpm-rs/tpm-rs#61]
It would be awesome for debugging to decode RCs into nice human-readable debug printing that details the error/layer/other metadata if present.

**Comment by a-pronin** [comment]
Yep, good idea, but this should be behind a compile-time feature to allow the embedded systems to avoid the cost of having this parsing/explanation logic in their code base.

**Comment by trent-reed** [comment]
I'll take this since I'm already doing error code refactoring. But also IDK that we need to hide behind a flag because unused functions are pruned by the final binary anyways. We can do testing to make sure that's the case - I don't want a feature option unless it's absolutely necessary - if unused functions are just dropped from the binary then I don't think we need a feature flag.

**Comment by oddcoder** [comment]
Those debug traits are exposed the outside of the compilation unit, so I don't think  they get automatically dead-code eliminated even on release builds, at least not everything? Experimentally the way I reduced the size of binary and eliminated unneeded parts of libraries was by adding the following to `cargo.toml`:

```toml
lto = true
codegen-units = 1
```

But I am not sure we should do this in tpm-rs since it is a library. That said I also slightly prefer not using feature flags for debug messages as well.

But I am curious, do you intend on using `core::fmt` or something else (ufmt, defmt, something else)?

**Comment by trent-reed** [comment]
The traits are definitely part of the crate, but I think you'll only keep what is used after linking into a binary is my point. I think rust has more thorough rules around dropping functions during the final linking phase than C or C++ (where it's all based on the translation unit).

However, I need to do more testing to be sure of this.

---

And yes, it's generally recommended to implement `Display` and `Debug` for error types, and if you have a more detailed display type you can use the adapter pattern and do something like this:

```rs
struct DetailedDisplay<'a>(&'a ErrorType);

struct ErrorType {
    ...
}

impl ErrorType {
    pub fn detailed(&self) -> DetailedDisplay {
        DetailedDisplay::new(self)
    }
}

impl Display for DetailedDisplay {
    ...
}
```

Then if someone wants a detailed error message:

```rs
println!("{}", error.detailed());
```

One place you can see a pattern like this in std is `Path` where `Display` is purposefully not implemented on the base type to make it clear to the caller that they are opting into lossy-conversion of the `OsStr`, and instead they say that there is a `.display()` function that can return a UTF8-displayable path by doing lossy conversion: https://doc.rust-lang.org/std/path/struct.Display.html

---

So if I wanted to completely separate the simple display from the complex display I'd start with something like that.

Anyways, I'm thinking something like that, but I'd have to do some testing to see what would work best.

**Comment by oddcoder** [comment]
> And yes, it's generally recommended to implement `Debug` and `Debug` for error types, and if you have a more detailed display type you can use the adapter pattern and do something like this:

So `Debug` and `Display` are convenient for `assert_*` and `unwrap()` in tests. But my concern is about pulling `core::fmt` into `services` if we want to use the same `Debug` and `Display` for logging.

I do not know if it is good idea to offer debugging/logging output in `services` in our particular case (display output about what is happening inside the TPM over UART or any other protocol).

If the answer is no, then `core::fmt` is fine, since it cannot end up on the chip anyways. But if the answer is yes, then  `core::fmt` can be problematic.

First `core::fmt` can panic on IO, which is [generally undesirable](https://github.com/orgs/tpm-rs/discussions/5) in embedded rust software. Second, there is folklore that `core::fmt` is advertised as not suitable for embedded systems. I only had one experience with it, where (among  other tricks) we had to to avoid using it to get the code to fit on the mcu. But unfortunately I don't have numbers.

I am only aware of two solutions targeted at replacing `core::fmt` for embedded systems
- [defmt](https://defmt.ferrous-systems.com/): Apparently this is feature-rich which might be the exact opposite of what we want.
- [ufmt](https://crates.io/crates/ufmt): Looks pretty bare-bone to me.

probably there are more on crates.io

Following the pattern in https://github.com/orgs/tpm-rs/discussions/5, I just shamelessly went and checked what they are doing at caliptra. It looks like they use a [custom fork of ufmt](https://github.com/chipsalliance/caliptra-sw/blob/51ff0a89f169bbf8e06acb49b31db555e99fefb6/Cargo.toml#L170) which they maintain. That is not to say we should blindly follow caliptra, but we might want to first consider the use-case of debug printing and pick a solution accordingly?

**Comment by trent-reed** [comment]
Those are some great points - my main goal is that there's some path for good debug output for the client, and that it is implemented in a predictable way. But I didn't think about the trouble it could cause in `service` since I was thinking pretty much solely about the client.

I'll prototype some things and play around a bit and make some suggestions, but it may in fact be the case that we will need a feature flag due to the reasons you have stated.

**Comment by oddcoder** [comment]
> I do not know if it is good idea to offer debugging/logging output in services in our particular case (display output about what is happening inside the TPM over UART or any other protocol).

I talked to some colleagues, and the general feedback I learned is the following:

- It will be hard for vendors to adapt tpm-rs to their platforms without having debugging capability in services for development boards.
- For final TPM products, the pins for debugging (like tx, rx in case of UART) should be disabled but also the debugging subsystem should be removed totally from the code in case an attacker found a way re-enable logging pins.

Aside from this discussion I looked into both `ufmt` and `demft` and the caliptra's fork of `ufmt`. `defmt` look totally not suitable for our usecase because of those two limitations:
- Output object format must be ELF
- Custom linking (linker script) is required

`ufmt` is more suitable but in my opinion it is not quite there yet. The maintainer stopped accepting patches for many years. I am not sure of the reason. Apparently an there are good ones like:
- [Get rid of UB with uninitialized](https://github.com/japaric/ufmt)
- [feat: implement `uDebug` for `PhantomData`](https://github.com/japaric/ufmt/pull/58)
- [Add "inline" feature for do_as_formatter()](https://github.com/japaric/ufmt/pull/59) which is the caliptra fork.

Additionally given our special usecase, we might want for `ufmt` to be drop in replaced by `core::fmt` when we are outside `services`, or when we are inside services but linking against `std` (swtpm target for instance). There are some ideas in [this issue in [rustlang/log](https://github.com/rust-lang/log/issues/332). I have some possible ideas as well but I want to know what you think?
One more thing, while searching, I found this [benchmark](https://docs.rs/ufmt/latest/ufmt/index.html#benchmarks). It might prove useful.
